tensorboard/webapp/widgets/custom_modal/custom_modal.ts

```diff
@@ -0,0 +1,221 @@
+export class CustomModal {
+  private customModalRefs: CustomModalRef[] = [];
+
+  constructor(
+    private readonly appRef: ApplicationRef,
+    private readonly overlay: Overlay
+  ) {}
+
+  /** Gets the ViewContainerRef of the app's root component if it exists. */
+  private getModalViewContainerRef(): ViewContainerRef | undefined {
+    const appComponents = this.appRef.components;
+    if (appComponents.length === 0) {
+      // appComponents can potentially be empty in tests.
+      return;
+    }
+
+    const appInstance = appComponents[0].instance;
+    let viewContainerRef: ViewContainerRef = appInstance.vcRef;
+    if (!viewContainerRef) {
+      console.warn(
+        'For proper custom modal function, an ViewContainerRef named `vcRef` is required in the root component.'
+      );
+      return;
+    }
+    return viewContainerRef;
+  }
+
+  /** Creates a modal from a template next to an element. */
+  createNextToElement(
+    templateRef: TemplateRef<unknown>,
+    element: Element,
+    connectedPosition: ConnectedPosition = {
+      originX: 'end',
+      originY: 'top',
+      overlayX: 'start',
+      overlayY: 'top',
+    }
+  ): CustomModalRef | undefined {
+    const viewContainerRef = this.getModalViewContainerRef();
+    if (!viewContainerRef) return;
+
+    let positionStrategy = this.overlay.position().flexibleConnectedTo(element);
+    if (connectedPosition) {
+      positionStrategy = positionStrategy.withPositions([connectedPosition]);
+    }
+
+    const overlayRef = this.overlay.create({
+      positionStrategy,
+      hasBackdrop: false,
+    });
+    overlayRef.attach(new TemplatePortal(templateRef, viewContainerRef));
+    const customModalRef = new CustomModalRef(overlayRef);
+    this.customModalRefs.push(customModalRef);
+
+    // setTimeout to prevent closing immediately after modal open.
+    setTimeout(() => {
+      const outsidePointerEventsSubscription = overlayRef
+        .outsidePointerEvents()
+        .subscribe((event) => {
+          // Only close when click is outside of every modal
+          if (
+            this.customModalRefs.every(
+              (ref) =>
+                !isMouseEventInElement(event, ref.overlayRef.overlayElement)
+            )
+          ) {
+            this.closeAll();
+          }
+        });
+      customModalRef.subscriptions.push(outsidePointerEventsSubscription);
+    });
+
+    const keydownEventsSubscription = overlayRef
+      .keydownEvents()
+      .subscribe((event) => {
+        if (event.key === 'Escape') {
+          this.closeAll();
+        }
+      });
+    customModalRef.subscriptions.push(keydownEventsSubscription);
+
+    return customModalRef;
+  }
+
+  /** Destroys given custom modal and related resources. */
+  close(customModalRef: CustomModalRef) {
+    const index = this.customModalRefs.findIndex(
+      (ref) => ref === customModalRef
+    );
+    if (index === -1) {
+      console.warn('Could not find customModalRef', customModalRef);
```

Comment:
I got this message many times. I think mainly from choosing the "Filter" submenu from the context menu?

Reply:
This was from a dangling customModalRef pointer in DataTableComponent: if CustomModal.closeAll() is called, the modals refs are cleaned up in CustomModal, but not in DataTableComponent, and the next DataTableComponent.closeSubmenus() will request closing a nonexistent modal.

I added an onClose Subject to CustomModalRef that callers can subscribe to and run cleanup code, like `this.filterModalRef = undefined;` (also added an onOpen Subject for symmetry)